python/dgl/contrib/graph_store.py

```diff
@@ -0,0 +1,355 @@
+    """
+    __slots__ = ['_graph', '_edges', '_graph_name']
+
+    def __init__(self, graph, edges, graph_name):
+        self._graph = graph
+        self._edges = edges
+        self._graph_name = graph_name
+
+    def __getitem__(self, key):
+        return self._graph.get_e_repr(self._edges)[key]
+
+    def __setitem__(self, key, val):
+        # Move the data in val to shared memory.
+        dlpack = F.zerocopy_to_dlpack(val)
+        dgl_tensor = nd.from_dlpack(dlpack)
+        val = _CAPI_DGLCreateSharedMemWithData(_get_edata_path(self._graph_name, key),
+                                                dgl_tensor)
+        self._graph.set_e_repr({key : val}, self._edges)
+
+    def __delitem__(self, key):
+        if not is_all(self._edges):
+            raise DGLError('Delete feature data is not supported on only a subset'
+                           ' of nodes. Please use `del G.edata[key]` instead.')
+        self._graph.pop_e_repr(key)
+
+    def __len__(self):
+        return len(self._graph._edge_frame)
+
+    def __iter__(self):
+        return iter(self._graph._edge_frame)
+
+    def __repr__(self):
+        data = self._graph.get_e_repr(self._edges)
+        return repr({key : data[key] for key in self._graph._edge_frame})
+
+def _to_csr(graph_data, edge_dir, multigraph):
+    try:
+        indptr = graph_data.indptr
+        indices = graph_data.indices
+        return indptr, indices
+    except:
+        if isinstance(graph_data, scipy.sparse.spmatrix):
+            csr = graph_data.tocsr()
+            return csr.indptr, csr.indices
+        else:
+            idx = create_graph_index(graph_data=graph_data, multigraph=multigraph, readonly=True)
+            transpose = False if edge_dir == 'in' else True
+            csr = idx.adjacency_matrix_scipy(transpose, 'csr')
+            return csr.indptr, csr.indices
+
+class SharedMemoryStoreServer:
+    """The graph store server.
+
+    The server loads graph structure and node embeddings and edge embeddings
+    and store them in shared memory.
+
+    Parameters
+    ----------
+    graph_data : graph data
+        Data to initialize graph. Same as networkx's semantics.
+    edge_dir : string
+        the edge direction for the graph structure.
+    graph_name : string
+        Define the name of the graph.
+    multigraph : bool, optional
+        Whether the graph would be a multigraph (default: False)
+    num_workers : int
+        The number of workers that will connect to the server.
+    port : int
+        The port that the server listens to.
+    """
+    def __init__(self, graph_data, edge_dir, graph_name, multigraph, num_workers, port):
+        graph_idx = GraphIndex(multigraph=multigraph, readonly=True)
+        indptr, indices = _to_csr(graph_data, edge_dir, multigraph)
+        graph_idx.from_csr_matrix(indptr, indices, edge_dir, _get_graph_path(graph_name))
+
+        self._graph = DGLGraph(graph_idx, multigraph=multigraph, readonly=False)
```

Comment:
So we first convert the graph data to CSR, create an *immutable* graph index and initialize from CSR, and then initialize a *mutable* `DGLGraph` atop that?

Reply:
yes

Reply:
Why do we want a mutable DGLGraph atop an immutable index then?

Reply:
it's a bug. it has been fixed.